Add an open/close button for the Sidebar

julienw opened the pull request:
* Closed:
![image](https://user-images.githubusercontent.com/454175/42104227-154ff4a6-7bcc-11e8-8ef0-00410a856672.png)

* Open:
![image](https://user-images.githubusercontent.com/454175/42104210-0860f704-7bcc-11e8-9245-b6c639a1345b.png)


With Harald we discussed that it should be open by default, but I made it closed by default for now while we finish polishing. By being closed by default, I think we can enable it once #1041 lands (without that PR it's quite useless).

[Here's a direct link to this deployed PR with a profile](https://deploy-preview-1096--perf-html.netlify.com/public/280e10708c0660695ef5af145c53179d29d9796c/calltree/?thread=4&threadOrder=0-2-3-4-5-1&v=3).

julienw commented on res/css/style.css:
```diff
@@ -602,20 +602,20 @@ body, #root, .profileViewer {
 }
 
 .tabBarTabWrapper {
-  margin: 0;
```
overriden below

julienw commented on res/css/style.css:
```diff
@@ -602,20 +602,20 @@ body, #root, .profileViewer {
 .tabBarContainer {
   display: flex;
   flex-flow: row nowrap;
+  justify-content: space-between;
```
so that the "sidebar" button is at the right while the tab bar is at the left. (maybe I should add this as a comment ?)

julienw replied:
OK I'll add it.

julienw commented on src/components/sidebar/index.js:
```diff
@@ -15,8 +15,8 @@ export default function selectSidebar(
-    //calltree: CallTreeSidebar,
-    calltree: null,
+    calltree: CallTreeSidebar,
+    //calltree: null,
```
I won't commit this, but this makes it easier to test.

zoepage commented on res/css/photon-components.css:
```diff
@@ -0,0 +1,34 @@
+  border: none;
+}
+
+/* Using -moz-focusring to target Firefox only, as a result of the style above */
```
Nice catch :)

julienw replied:
I have no merit => https://fvsch.com/code/button-css/ :)

julienw commented on res/css/photon-components.css:
```diff
@@ -0,0 +1,34 @@
+
+  /* photon styles */
+  padding: 0 8px;
+  border-radius: 2px;
```
ah I see what you mean, good feedback

julienw commented on res/css/style.css:
```diff
@@ -633,6 +635,8 @@ body, #root, .profileViewer {
   border: solid transparent;
   border-width: 0 1px 0 1px;
+  overflow: hidden;
+  text-overflow: ellipsis;
```
For the shrink behavior

julienw commented on res/css/style.css:
```diff
@@ -602,20 +603,21 @@ body, #root, .profileViewer {
   flex-flow: row nowrap;
   margin: 0 -1px;
-  max-width: 100%;
+  min-width: 0; /* This makes the tab container actually shrinkable below min-content */
```
Eventually we should move these styles to a separate CSS file and reorder them a bit...

zoepage replied:
Might be worth a look, yes. Wanna open an issue for this?

julienw replied:
Sure, I filed #1119

julienw commented:
@zoepage it's ready for a full review !
@violasong I used the devtools icon for now. You can see screenshots above and try it out using the link in the description above

violasong commented:
This looks great - thanks Julien!

julienw commented:
Note: I tried to split the PR in small commits, but please tell me if it's too big and I can work on splitting it in several PRs instead.

julienw commented:
A quick note that I forgot something when I did this: we want a boolean state _per panel_. I'll work on this later today, but in the mean time please do not review... Sorry about this.

zoepage commented:
What's the status here @julienw? Just want to make sure to plan in time to review as soon as possible, so we can merge. :)

julienw commented:
I think it will be ready for a review later today. I'm nearly there, but I needed #1112 before so that I can properly initialize the state here.

julienw commented on src/app-logic/tabs-handling.js:
```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This is part of #1112.

julienw commented on src/components/sidebar/index.js:
```diff
@@ -5,19 +5,19 @@
 
```
As usual, the changes in this file won't be merged in this PR. After approval I'll remove the commit with this change and if necessary regenerate the various test snapshots that are affected.

I enabled the sidebar in 2 different panels so that we can test that the state is per panel.

julienw commented on src/reducers/app.js:
```diff
@@ -63,10 +72,37 @@ function hasZoomedViaMousewheel(state: boolean = false, action: Action) {
+  if (state === undefined) {
+    state = {};
+    tabSlugs.forEach(tabSlug => (state[tabSlug] = false));
+  }
```
Because the computation of the default state isn't simple, I decided to do it here instead of using the default arguments like elsewhere.

zoepage replied:
I'd say that is totally okay until we find a cleaner solution :)

julienw commented on src/components/app/Details.js:
```diff
@@ -79,16 +63,40 @@ class ProfileViewer extends PureComponent<Props> {
           tabOrder={tabOrder}
           onSelectTab={this._onSelectTab}
           onChangeTabOrder={changeTabOrder}
+          extraElements={extraButton}
```
I went back and forth about how to insert the button in the DOM. In the end I decided to include it in the TabBar, but still define it in `Details`, because I thought that while it was part of the `TabBar` visually the TabBar itself shouldn't have to know anything about the sidebar.

julienw commented:
@zoepage I think it's now ready ! As usual I tried to split the PR in individual easier to review commits.
Github doesn't show the commits in the "right" order, so let me just say which commits you don't need to look in this PR:
* `Extract tabs list in a separate file`, as this is #1112. Note that this commit touches a lot of files and this bumps the number of changed files quite a bit.
* `Enable sidebar` as this is only for testing purpose and I'll remove it before merging

`Add Redux state for the open/closed state of the sidebar` that's displayed between these commits _is_ part of this PR though.

Please ask if you have any question about something here :)

julienw commented on src/components/app/Details.js:
```diff
@@ -61,6 +96,7 @@ class ProfileViewer extends PureComponent<Props> {
           tabOrder={tabOrder}
           onSelectTab={this._onSelectTab}
           onChangeTabOrder={changeTabOrder}
+          extraElements={extraButton}
```
I went back and forth about how to insert the button in the DOM. In the end I decided to include it in the TabBar, but still define it in Details, because I thought that while it was part of the TabBar visually the TabBar itself shouldn't have to know anything about the sidebar.

zoepage replied:
Sounds reasonable.

zoepage commented on src/components/app/DetailsContainer.js:
```diff
@@ -24,10 +25,11 @@ function dispatchResizeEvent() {
-function DetailsContainer({ selectedTab }: StateProps) {
-  const Sidebar = selectSidebar(selectedTab);
+function DetailsContainer({ selectedTab, isSidebarOpen }: StateProps) {
+  const Sidebar = isSidebarOpen && selectSidebar(selectedTab);
```
Very small nit,
can we switch the order of `selectSidebar(selectedTab)`  and `isSidebarOpen` here?

1. `mapStateToProps` and the params both are in that order
2. As `isSidebarOpen` depends on `selectSidebar(selectedTab)`, it would fail first and be more fault tolerant.

julienw replied:
I don't understand this comment @zoepage. Especially `As isSidebarOpen depends on selectSidebar(selectedTab)`, what do you mean here ? I also don't understand the `fault tolerant` part.

The function `selectSidebar` returns the component I'll use as the Sidebar, I can't put it first otherwise I'll lose the result.
Moreover in a shortcircuit construct like `&&` it's better for performance to put the less costly operation first (in this case a boolean is less costly than a function call). Here the performance is not important because `selectSidebar` is very cheap too.

julienw commented:
I added the comment, I removed the commit that enables the sidebar, and I regenerated the snapshots that failed because of this.

Then I'll use my adminitrator power to merge following Ola's approval on Slack.
